File: graph-service-examples/python/load_balancer/load_balancer.py
# Copyright 2022-2026 Aerospike, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from aiohttp import ClientConnectorError, ServerDisconnectedError
from gremlin_python.driver.driver_remote_connection import DriverRemoteConnection
import gc
import threading
import logging

from gremlin_python.driver.resultset import ResultSet
from gremlin_python.process.traversal import Bytecode


class RoundRobinClientRemoteConnection:
    def __init__(self, endpoints, traversal_source="g", health_check_interval=10,
                 logger: logging.Logger = None, log_level: int = logging.INFO):

        self._logger = logger or logging.getLogger(self.__class__.__name__)
        self._logger.setLevel(log_level)

        self._clients = [
            DriverRemoteConnection(f"ws://{host}/gremlin", traversal_source)
            for host in endpoints
        ]
        self._traversal_source = traversal_source
        self._available = [True] * len(self._clients)
        self._pos = 0
        self._lock = threading.Lock()

        # health checker
        self._health_interval = health_check_interval
        self._stop_event = threading.Event()
        self._health_thread   = threading.Thread(
            target=self.health_check_loop, daemon=True
        )
        self._health_thread.start()

        self._logger.debug("Initialized load-balancer with endpoints: %s", endpoints)

    def add_host(self, endpoint):
        self._clients.append(DriverRemoteConnection(f"ws://{endpoint}/gremlin", self._traversal_source))
        self._available.append(True)
        self._logger.info("Added host %s", endpoint)

    def remove_host(self, endpoint):
        try:
            with self._lock:
                index = -1
                for host in self._clients:
                    if endpoint in host.url:
                        index = self._clients.index(host)
                if not index == -1:
                    del self._clients[index]
                    del self._available[index]
                    self._logger.info("Removed host %s", endpoint)
                else:
                    self._logger.warning("Tried to remove non-existent host %s", endpoint)
        except ValueError:
            self._logger.warning("Tried to remove non-existent host %s", endpoint)

    def submit(self, bytecode: Bytecode, aliases=None) -> ResultSet:
        for _ in range(len(self._clients)):
            with self._lock:
                healthy = [i for i, ok in enumerate(self._available) if ok]
                if not healthy:
                    raise RuntimeError("No healthy Gremlin hosts available")
                pick = healthy[self._pos % len(healthy)]
                self._pos += 1
            try:
                result = self._clients[pick].submit(bytecode)
                with self._lock:
                    self._available[pick] = True
                self._logger.debug("Traversal submitted via connection #%d", pick)
                return result
            except (ClientConnectorError, ServerDisconnectedError) as e:
                with self._lock:
                    self._available[pick] = False
                self._logger.warning("Connection #%d failed: %s – marking host down", pick, e)
                last_exc = e
            except Exception:
                raise
        self._logger.error("All endpoints failed – raising")
        raise RuntimeError("All Gremlin endpoints failed") from last_exc

    def health_check_loop(self):
        while not self._stop_event.wait(self._health_interval):
            self._logger.debug("Running health check")
            for i, ok in enumerate(self._available):
                if not ok:
                    try:
                        self._clients[i]._client.submit("g.V().limit(1).toList()")
                        with self._lock:
                            self._available[i] = True
                            self._logger.info("Host #%d is healthy again", i)
                    except Exception:
                        with self._lock:
                            self._logger.debug("Host #%d still down", i)
                            self._available[i] = False
                        pass

    def get_clients(self):
        return list(self._clients)

    def get_available(self):
        return list(self._available)

    def close(self):
        self._stop_event.set()
        self._health_thread.join()
        # Force collection of any lingering aiohttp ClientResponse objects before
        # each connection's private event loop closes, avoiding "Event loop is
        # closed" errors from __del__ finalizers running after the fact.
        gc.collect()
        for c in self._clients:
            try:
                c.close()
            except Exception:
                pass
        gc.collect()
        self._logger.debug("Load-balancer shut down")

File: graph-service-examples/nodejs/load-balancer/load-balancer.js
const gremlin = require('gremlin');
const { DriverRemoteConnection } = gremlin.driver;
const { traversal } = gremlin.process;

class RoundRobinClientRemoteConnection {
    constructor(
        endpoints,
        traversalSource = 'g',
        healthCheckIntervalSec = 10,
        logger = null
    ) {
        this._traversalSource = traversalSource;
        this._logger =
            logger ||
            ((level, msg) => {
                if (level === 'debug') console.debug(msg);
                else if (level === 'info') console.info(msg);
                else if (level === 'warn') console.warn(msg);
                else console.log(msg);
            });

        this._clients = endpoints.map(
            (host) => new DriverRemoteConnection(`ws://${host}/gremlin`, { traversalSource })
        );
        this._available = Array(this._clients.length).fill(true);
        this._pos = 0;

        this._stopped = false;
        this._healthIntervalMs = Math.max(1, healthCheckIntervalSec) * 1000;
        this._healthTimer = setInterval(() => this._healthCheckLoop(), this._healthIntervalMs);

        this._debug(`Initialized load-balancer with endpoints: ${endpoints.join(', ')}`);
    }

    async submit(bytecode, requestOptions) {
        let lastErr = null;

        for (let attempt = 0; attempt < this._clients.length; attempt++) {
            const healthy = [];
            for (let i = 0; i < this._available.length; i++) {
                if (this._clients[i] && this._available[i]) healthy.push(i);
            }
            if (healthy.length === 0) {
                throw new Error('No healthy Gremlin hosts available');
            }
            const pick = healthy[this._pos % healthy.length];
            this._pos += 1;

            const pickedConn = this._clients[pick];

            try {
                const rs = await pickedConn.submit(bytecode, requestOptions);
                this._available[pick] = true;
                this._debug(`Traversal submitted via connection #${pick}`);
                return rs;
            } catch (e) {
                const idxNow = this._clients.indexOf(pickedConn);
                if (idxNow !== -1) this._available[idxNow] = false;
                this._warn(`Connection #${idxNow !== -1 ? idxNow : '?'} failed: ${e?.message || e} – marking host down`);
                lastErr = e;
            }
        }

        this._error('All endpoints failed – raising');
        const err = new Error('All Gremlin endpoints failed');
        err.cause = lastErr;
        throw err;
    }

    async close() {
        if (this._stopped) return;
        clearInterval(this._healthTimer);
        this._stopped = true;

        const tasks = [];
        for (const c of this._clients) {
            if (!c) continue;
            try {
                tasks.push(c.close().catch(() => {}));
            } catch {}
        }
        await Promise.allSettled(tasks);
        this._debug('Load balancer closed');
    }

    addHost(endpoint) {
        const c = new DriverRemoteConnection(`ws://${endpoint}/gremlin`, {
            traversalSource: this._traversalSource,
        });
        this._clients.push(c);
        this._available.push(true);
        this._info(`Added host ${endpoint}`);
    }

    removeHost(endpoint) {
        let removed = false;
        for (let i = this._clients.length - 1; i >= 0; i--) {
            const c = this._clients[i];
            const url = c && (c.url || c._url || '');
            if (c && url.includes(endpoint)) {
                try { c.close().catch(() => {}); } catch {}
                this._clients[i] = null;
                this._available[i] = false;
                removed = true;
            }
        }
        if (removed) this._info(`Removed host ${endpoint}`); else this._warn(`Tried to remove non-existent host ${endpoint}`);
    }

    getClients() {
        return this._clients.filter(Boolean);
    }

    getAvailable() {
        const out = [];
        for (let i = 0; i < this._clients.length; i++) {
            if (this._clients[i]) out.push(!!this._available[i]);
        }
        return out;
    }

    async _healthCheckLoop() {
        if (this._stopped) return;
        this._debug('Running health check');
        const tasks = [];

        for (let i = 0; i < this._clients.length; i++) {
            const conn = this._clients[i];
            if (!conn) continue;
            if (this._available[i]) continue;
            tasks.push(
                (async () => {
                    try {
                        const g = traversal().withRemote(conn);
                        await g.V().limit(1).toList();
                        this._available[i] = true;
                        this._info(`Host #${i} is healthy again`);
                    } catch {
                        this._available[i] = false;
                        this._debug(`Host #${i} still down`);
                    }
                })()
            );
        }
        await Promise.allSettled(tasks);
    }

    _debug(msg) {
        this._logger('debug', msg);
    }
    _info(msg) {
        this._logger('info', msg);
    }
    _warn(msg) {
        this._logger('warn', msg);
    }
    _error(msg) {
        this._logger('error', msg);
    }
}

module.exports = { RoundRobinClientRemoteConnection };
